File: packages/snow_draw_engine/lib/draw/types/draw_color.dart
import 'package:meta/meta.dart';

/// Backend-agnostic color value stored as packed ARGB32 (`0xAARRGGBB`).
@immutable
final class DrawColor {
  /// Creates a color from packed ARGB32 (`0xAARRGGBB`).
  const DrawColor(int argb32) : argb32 = argb32 & 0xFFFFFFFF;

  /// Packed ARGB32 value (`0xAARRGGBB`).
  final int argb32;

  /// Alpha channel in `[0, 255]`.
  int get alpha => (argb32 >>> 24) & 0xFF;

  /// Red channel in `[0, 255]`.
  int get red => (argb32 >>> 16) & 0xFF;

  /// Green channel in `[0, 255]`.
  int get green => (argb32 >>> 8) & 0xFF;

  /// Blue channel in `[0, 255]`.
  int get blue => argb32 & 0xFF;

  /// Normalized alpha in `[0, 1]`.
  double get a => alpha / 255;

  /// Normalized red in `[0, 1]`.
  double get r => red / 255;

  /// Normalized green in `[0, 1]`.
  double get g => green / 255;

  /// Normalized blue in `[0, 1]`.
  double get b => blue / 255;

  /// Returns this color as packed ARGB32 (`0xAARRGGBB`).
  int toARGB32() => argb32;

  /// Returns a copy with [alpha] (0-255).
  DrawColor withAlpha(int alpha) {
    final normalizedAlpha = alpha.clamp(0, 255);
    return DrawColor((normalizedAlpha << 24) | (argb32 & 0x00FFFFFF));
  }

  /// Returns a copy with optional normalized channels in `[0, 1]`.
  ///
  /// This mirrors Flutter's `Color.withValues` API for alpha-driven updates
  /// used in reducers/encoders during migration.
  DrawColor withValues({
    double? alpha,
    double? red,
    double? green,
    double? blue,
  }) {
    int resolve(double? value, int fallback) {
      if (value == null) {
        return fallback;
      }
      final normalized = value.clamp(0.0, 1.0);
      return (normalized * 255).round().clamp(0, 255);
    }

    final nextAlpha = resolve(alpha, this.alpha);
    final nextRed = resolve(red, this.red);
    final nextGreen = resolve(green, this.green);
    final nextBlue = resolve(blue, this.blue);
    return DrawColor(
      (nextAlpha << 24) | (nextRed << 16) | (nextGreen << 8) | nextBlue,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is DrawColor && other.argb32 == argb32;

  @override
  int get hashCode => argb32.hashCode;

  @override
  String toString() =>
      'DrawColor(0x${argb32.toRadixString(16).padLeft(8, '0')})';
}

File: packages/snow_draw_rust_engine/src/draw/types/draw_color.rs
#![allow(dead_code)]

use serde::{Deserialize, Serialize};
use std::fmt;

/// Backend-agnostic color value stored as packed ARGB32 (`0xAARRGGBB`).
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct DrawColor {
    /// Packed ARGB32 value (`0xAARRGGBB`).
    pub argb32: u32,
}

impl DrawColor {
    /// Creates a color from packed ARGB32 (`0xAARRGGBB`).
    pub const fn new(argb32: u32) -> Self {
        Self { argb32 }
    }

    /// Alpha channel in `[0, 255]`.
    pub const fn alpha(self) -> u8 {
        ((self.argb32 >> 24) & 0xFF) as u8
    }

    /// Red channel in `[0, 255]`.
    pub const fn red(self) -> u8 {
        ((self.argb32 >> 16) & 0xFF) as u8
    }

    /// Green channel in `[0, 255]`.
    pub const fn green(self) -> u8 {
        ((self.argb32 >> 8) & 0xFF) as u8
    }

    /// Blue channel in `[0, 255]`.
    pub const fn blue(self) -> u8 {
        (self.argb32 & 0xFF) as u8
    }

    /// Normalized alpha in `[0, 1]`.
    pub fn a(self) -> f64 {
        self.alpha() as f64 / 255.0
    }

    /// Normalized red in `[0, 1]`.
    pub fn r(self) -> f64 {
        self.red() as f64 / 255.0
    }

    /// Normalized green in `[0, 1]`.
    pub fn g(self) -> f64 {
        self.green() as f64 / 255.0
    }

    /// Normalized blue in `[0, 1]`.
    pub fn b(self) -> f64 {
        self.blue() as f64 / 255.0
    }

    /// Returns this color as packed ARGB32 (`0xAARRGGBB`).
    pub const fn to_argb32(self) -> u32 {
        self.argb32
    }

    /// Returns a copy with `alpha` (`0..=255`).
    pub fn with_alpha(self, alpha: i32) -> Self {
        let normalized_alpha = alpha.clamp(0, 255) as u32;
        Self::new((normalized_alpha << 24) | (self.argb32 & 0x00FF_FFFF))
    }

    /// Returns a copy with optional normalized channels in `[0, 1]`.
    pub fn with_values(
        self,
        alpha: Option<f64>,
        red: Option<f64>,
        green: Option<f64>,
        blue: Option<f64>,
    ) -> Self {
        let next_alpha = resolve_channel(alpha, self.alpha());
        let next_red = resolve_channel(red, self.red());
        let next_green = resolve_channel(green, self.green());
        let next_blue = resolve_channel(blue, self.blue());

        Self::new((next_alpha << 24) | (next_red << 16) | (next_green << 8) | next_blue)
    }
}

impl fmt::Display for DrawColor {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "DrawColor(0x{:08x})", self.argb32)
    }
}

fn resolve_channel(value: Option<f64>, fallback: u8) -> u32 {
    match value {
        None => fallback as u32,
        Some(channel) => {
            if !channel.is_finite() {
                return fallback as u32;
            }
            (channel.clamp(0.0, 1.0) * 255.0).round().clamp(0.0, 255.0) as u32
        }
    }
}
